servers/physics_server_2d.cpp

```diff
@@ -132,9 +132,14 @@ PhysicsDirectBodyState2D::PhysicsDirectBodyState2D() {}
 
 ///////////////////////////////////////////////////////
 
-void PhysicsShapeQueryParameters2D::set_shape(const RES &p_shape) {
-	ERR_FAIL_COND(p_shape.is_null());
-	shape = p_shape->get_rid();
+void PhysicsShapeQueryParameters2D::set_shape(const RES &p_shape_ref) {
+	ERR_FAIL_COND(p_shape_ref.is_null());
+	shape_ref = p_shape_ref;
+	shape = p_shape_ref->get_rid();
+}
+
+RES PhysicsShapeQueryParameters2D::get_shape() const {
+	return shape_ref;
 }
 
 void PhysicsShapeQueryParameters2D::set_shape_rid(const RID &p_shape) {
```

Comment:
`shape_ref` should be reset here to keep things consistent.

Reply:
By resetting you mean this?

```suggestion
void PhysicsShapeQueryParameters2D::set_shape_rid(const RID &p_shape) {
	shape = p_shape;
	shape_ref = RES();
```

In that case we don't want to lose the actual reference if the `p_shape == shape`, will add an additional check if possible.

Reply:
Yes that's what I meant! And I agree, it's safer to first check if there's a stored shape and if the rid is different.